src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -138,6 +95,165 @@ void GcsNodeManager::ScheduleTick() {
   });
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+GcsNodeManager::GcsNodeManager(boost::asio::io_service &io_service,
+                               gcs::NodeInfoAccessor &node_info_accessor,
+                               gcs::ErrorInfoAccessor &error_info_accessor)
+    : node_info_accessor_(node_info_accessor),
+      error_info_accessor_(error_info_accessor),
+      node_failure_detector_(new NodeFailureDetector(
+          io_service, node_info_accessor, [this](const ClientID &node_id) {
+            if (auto node = RemoveNode(node_id, /* is_intended = */ false)) {
+              node->set_state(rpc::GcsNodeInfo::DEAD);
+              RAY_CHECK(dead_nodes_.emplace(node_id, node).second);
+              RAY_CHECK_OK(node_info_accessor_.AsyncUnregister(node_id, nullptr));
+              // TODO(Shanly): Remove node resources from resource table.
+            }
+          })) {
+  // TODO(Shanly): Load node info list from storage synchronously.
+  const auto lookup_callback =
+      [this](Status status, const std::vector<rpc::GcsNodeInfo> &node_info_list) {
+        for (const auto &node_info : node_info_list) {
+          if (node_info.state() != rpc::GcsNodeInfo::DEAD) {
+            AddNode(std::make_shared<rpc::GcsNodeInfo>(node_info));
+          }
```

Comment:
Also need to put dead nodes in the `dead_nodes_` map?

Reply:
> Also need to put dead nodes in the `dead_nodes_` map?

Yes, u r right. I will fix it soon.